### PR TITLE
allow configuring imlib2's cache size

### PR DESCRIPTION
**Pull request by N-R-K**
by default imlib2 uses a 4mb cache, which is quite small. this allows
users who have more memory to spare to set a bigger cache size and avoid
reloading an already viewed image if it fits into the cache.

### CODE REVIEW

**Comment by TAAPArthur** [comment]
I don't really mind this change, but it seems like something that should be trivial to do from the config file since one would just need to call a pretty simple function. I question if the fact that it isn't trivial a larger problem.

Let me phrase it differently: Are there other simple tweaks users may want to do? Would a general `on_start` function help solve other trivial problems. Should this be a cli option so it can be changed at runtime? 

And to answer my own questions... `on_start` seems too powerful and better left to a patch. A cli option seems more useful albeit more inconvenient if the general usecase is to never change and can't off hand would think of a reason why one would want to change it at runtime outside of binary distros.

Honestly though does the disk latency really matter? I can imagine it would on a remote FS, but I can't imagine that is the norm. Curious if you benefit from this are are just doing it because it seems like a good setting to have

**Comment by N-R-K** [comment]
> Would a general on_start function help solve other trivial problems.

Adding a general `on_start` function opens up a bag of worms. What if someone demands a `on_win_init` or `on_img_close` or `on_parse_option` etc etc. I don't exactly hate the idea of having some hooks though, but I can also see how it can get out of hand.

> Curious if you benefit from this are are just doing it because it seems like a good setting to have

Not really, I was messing around with cache size while trying to investigate #70 . Most of the time, it didn't matter. But there was one specific image where I noticed having a larger cache noticeably helped.

[Here's the image in question](https://postimg.cc/n914SfDN/9c08a597), it blows up the cache (around 200mb) when decoded/uncompressed and takes a noticeable amount of time to open (~2sec). With the default cache size, everytime you go back/previous the entire image needs to be reloaded and you suffer a ~2sec penalty.

I don't benefit from this patch much since all (except this one image) I have aren't that big to blow up the cache and have slow load time. But I assume this would be helpful for people dealing with large images.

**Comment by N-R-K** [comment]
Hi @0ion9 , giving a ping since I remember you mentioning working with large image files. Have you tried messing around with the Imlib2 cache size? If you have was it useful or no?

**Comment by 0ion9** [comment]
oh, thanks. I haven't tried since I didn't know it was relevant (re the reloading issue I have mentioned, shouldn't any cached data be strictly ignored by definition, though?). 

Considering the sizes I am working with I can guess that a 60mb or more cache could help (A4 at 450dpi == 19.6M pixels -> 58.8mb if it's stored as 24bit rgb, 73.6mb if stored as 32bit rgba; I also have some A3s which are 2x the size). Maybe it should be double that so I can quickly do a 2-image 'flip back and forth'. I'm not sure. From my research the cache doesn't do anything too exotic,  just  'remember whole images associated with paths'.Correct?

I'll have a look at this patch and try some different numbers.

**Comment by N-R-K** [comment]
For the reload issue, yes the cache shouldn't help. But it should have noticeable impact when going next/previous images when dealing with large images. Try it out and let us know how it goes.

**Review by XPhyro — APPROVED** [review]
> [Here's the image in question](https://i.postimg.cc/sxkTXVDf/based-pepe.png), it blows up the cache (around 200mb) when decoded/uncompressed and takes a noticeable amount of time to open (~2sec). With the default cache size, everytime you go back/previous the entire image needs to be reloaded and you suffer a ~2sec penalty.

I can see a speedup with even 5MiB images (after bumping the cache size further). Opening it for the first time still takes the same time (~1sec), but returning to it is instantaneous.

> it seems like something that should be trivial to do from the config file since one would just need to call a pretty simple function.

Though it is trivial to implement, I think this should be merged. I doubt most people would know that such an option exists in imlib2 or know that their problem with a large image may be due to cache size to research it. Similarly, they may not see the patch. Some users will also not bother patching as they do not deal with massive images, but those few milliseconds for smaller images do add up.

Furthermore, it does not change anything by default, it just exposes a configuration variable to change the default. It's harmless, maintainable and can prove beneficial as it is quite common to see images of sizes in the MiB range. For instance, 22% of the images in this repo are strictly larger than 1MiB: [Anime-Girls-Holding-Programming-Books](https://github.com/laynH/Anime-Girls-Holding-Programming-Books)

I believe everybody can benefit from this whether they deal with small or large images, thus, it should not be left as a patch.

**Comment by N-R-K** [comment]
> For instance, 22% of the images in this repo are strictly larger than 1MiB

Just a small note, the filesize doesn't really have much to do with the cache. As far as I understand it's caching the decoded/uncompressed data. The pepe in question was for example ~3MB but it blew up to around 200MB when decoded.

**Comment by XPhyro** [comment]
> Just a small note, the filesize doesn't really have much to do with the cache. As far as I understand it's caching the decoded/uncompressed data.

In practice, I think it is safe to assume *some* correlation between the file size and the resulting data size.

**Comment by N-R-K** [comment]
I think it depends more on the resolution, compression being used, and how compression-friendly the image in question is.

**Comment by N-R-K** [comment]
Anyhow, the point you make about this being useful for a good majority of people stands.

Though it's only a config.h option now, so only source users will benefit. I don't think it makes sense to use a cli flag for this, xresources doesn't make much sense either since that's just for colors/fonts.

I've thought about bumping the default size to 16/32MB but that can make some users running on memory constrained systems angry.

**Comment by XPhyro** [comment]
Definitely, it is dependent on all factors. But I doubt the usual image will be compressed 6700%. The format of the image, the entropy in the image and all other factors do play a role, but in the end, you're compressing data. I'm not claiming this correlation is linear, but for the average image, there is *some* correlation.

---------

Nonetheless, the fact that we are caching the uncompressed raw data means that we will run out of cache even faster, meaning the 1MiB threshold I arbitrarily chose should be even lower. It also means that caching has even more impact on the speed as, we not only will not read the file again, but also will not decode/uncompress the contents.

I think we can bump the number to a conservative 8/16MiB by default. Those who are in an extreme memory-constrained situation can just build from the source. I do not expect many people are in a situation that cannot handle 8/16MiB.

**Comment by TAAPArthur** [comment]
My stance is that we shouldn't change the default behavior.

> Adding a general on_start function opens up a bag of worms. What if someone demands a on_win_init or on_img_close or on_parse_option etc etc. I don't exactly hate the idea of having some hooks though, but I can also see how it can get out of hand.

I generally agree and while it may be out of scope for this PR, it seems relevant to discuss since if we plan to have or at least open to hooks, may want to hold off of this change since it wouldn't be needed. And merging this now and reverting it later seems less than ideal.

>I doubt most people would know that such an option exists in imlib2 or know that their problem with a large image may be due to cache size to research it. Similarly, they may not see the patch. 

To be blunt, I think this is a terrible argument. I think we have assumed that the user base is willing to at least check this repo, nsxiv-extras and/or open an issue based on the fact that we deliberately don't support requested features because they can be done with shell scripts. I'm against your reasoning here more than your conclusion and agree that the simplicity of this PR makes it much more attractive than the stuff in extras.

>Furthermore, it does not change anything by default, it just exposes a configuration variable to change the default.

>I think we can bump the number to a conservative 8/16MiB by default. Those who are in an extreme memory-constrained situation can just build from the source. I do not expect many people are in a situation that cannot handle 8/16MiB.

Keeping the same value by default is why I don't have a strong opposition against this change. Increasing it by default seems really unneeded and out of line with being backwards compatible and portability. Also think we reached some concensus to try to keep sxiv behavior unless there is good justification not to. If you say users can build from source to get the previous behavior, then I think people who want to take advantage of this feature/config var should build from source to not affect existing users. I believe every other feature we've added has been opt-in and don't think this should be an exception.

>I don't think it makes sense to use a cli flag for this, xresources doesn't make much sense either since that's just for colors/fonts.

Doesn't make sense because it is so mundane or because people won't often want to change it? Currently xresources is just for fonts/colors but there isn't a rule enforcing that. I'm fine leaving it as a config option, but it seems about the same amount of effort (not counting docs) to stuff it into options and make it more accessible.

File: config.def.h
Comment (TAAPArthur) [inline config.def.h:63]:
```suggestion
static const int CACHE_SIZE = 4 << 20; /* 4MiB */
```

**Review by eylles — APPROVED** [review]

File: config.def.h
Comment (eylles) [reply config.def.h:63]: @N-R-K do u think this affects readabilty? 
I mean if anything this should perform better then again the operation is done only once so bitshift vs multiplication speed differnce is moot.

**Comment by 0ion9** [comment]
From my tests, the cache seems to be 32bit RGBA (at least for true color images). So a 4M-pixel image occupies 16Mib in memory, meaning that to get a quick flip between two such images the cache needs to be a little bigger than 32Mib.
For the images I mentioned, I've personally settled on 256Mib (~=3 A4 450dpi scans plus a little spare) 

The issue with GIFs seems to be ameliorated (less change in playback speed) but not solved by having a large cache. Also even if they, in theory , fit in cache simultaneously, there is still noticeable load time when switching images.

Based on the above testing, the default 4mb cache will be mostly filled by a 1024x768, and overfilled by a 1280x900 image.

File: config.def.h
Comment (TAAPArthur) [reply config.def.h:63]: I was aiming for readability. I don't think the generated object file changed since this is a pretty trivial optimization. I also think there are enough comments to make it clear what the value is in either case and leaned towards the more compact expression.

**Comment by mamg22** [comment]
I agree with bumping the cache size a little to 8/16MiB.

Out of curiosity, I've been looking at the source code history of Imlib2 searching for changes to the image cache size in the`cache_size` variable. It seems the 4MiB size has never changed since the [creation of Imlib2 in **1999**](https://git.enlightenment.org/legacy/imlib2.git/commit/?id=520edea668a1044786efc326dc85241a47600f9c), it might even be a value copied from the original Imlib.

File: config.def.h
Comment (N-R-K) [reply config.def.h:63]: It shouldn't matter performance wise, it will get optimized away. I think it's more clearer what's going on with `* 1024 * 1024` compared to bitshift.

**Comment by eylles** [comment]
If that was the cache back in 1999 then we may need to bump that value up consoderably for 2021, at least over 40MB i'd say

**Comment by N-R-K** [comment]
I think @TAAPArthur has a point about keeping the default though. If we want to follow the project scope, then we can't change the default. And if we want to make an exception, then we need clear rules for when to make exceptions. See: https://github.com/nsxiv/nsxiv/pull/173#issuecomment-978858667 for further discussion on that topic.

**Comment by eylles** [comment]
Welp, on this case i'd say there is a pretty convincing reason to bump up the default, for one it can provide a slightly better experience, for 2 nowadays we have way bigger images than there were around back in 1999, so i think we should scale the default cache to not get instantly filled with 1 image.

**Comment by N-R-K** [comment]
_If we are to bump the default cache size_ , a mere 16MB seems pointless. Just a simple 1440p image (not uncommon since a lot of smartphones support it) will almost fill it up.

It makes more sense to optimize for the most common case and go for either `256` . Or `128` if we want to be conservative, _maybe_ `64` if we want to be REALLY conservative. Anything below won't be all that useful and will just increase memory consumption with barely any performance improvement.

**Comment by mamg22** [comment]
Well, when put like that, I agree to follow the project scope, keep the current default and just let users pick whatever size they need.

**Comment by N-R-K** [comment]
Just some food for thought: Most desktop pc's have 8/16 gigs of memory. And even less resourceful systems like rpi have 4 gigs on average (according to a friend) and 1gigs on the lower end.

And even then, there's virtual memory, or more specifically disk swap, so it's pretty hard to actually OOM on any semi-modern system even with a low physical memory.

If we don't bump the size, then we know for sure that majority of the users will get lower image (re)loading performance. But so far, there hasn't been any ***practical example*** where bumping cache size to `64`/`128` would result in reduced portability.

Also, nsxiv is an image viewer running on X server. It isn't exactly some small cli tool. And we already _by default_ load animated images, which can potentially cause **far more** memory usage depending on how many frames there are on the image, since we're storing _all_ the frame on memory.

Loading [this 280x280 gif](https://i.postimg.cc/HxBZnrKX/real.gif) for example, causes nsxiv memory usage to go upto 75~mb for me. If someone really can't spare 64mb of memory then he'll have to compile from source anyways as otherwise opening a gif can potentially cause OOM.

**Comment by TAAPArthur** [comment]
To be honest talking about raising the default cache to be 8 to 50% of the ram of some systems seems like a non starter for a "minimal" image viewer.

Think about PinePhone with its 2GiB ram. Also keep in mind that having swap isn't mandatory and swapping (especially to slow sdcard) could kill perf and swapping happens before total memory is exhausted. For the case of gifs, not every user them (like me) or animated images in general. In other words they could have been using nsxiv previously and this could break them for what I view as little gain.

I hate the talk of practical examples because it boils down to what we (the people in this conversation) can imagine experiencing rather than what's possible. On my phone it seems very likely there would be sets of images that I could view that would work fine for sxiv but not nsxiv specifically because of this change.

Also Raspbery pi zeroes have .5 GiB in ram which is enough to use X + sxiv with small images. And technically it could be a remote X server so the memory cost of X doesn't have to matter.

>And even then, there's virtual memory, or more specifically disk swap, so it's pretty hard to actually OOM on any semi-modern system even with a low physical memory.

Simply not true. I don't know why we would want to make assumptions about hardware; there will be some example outside of expectations and we gain little by excluding them. I imagine the user base of minimal software has a above average rate of not using stock OS distributions / non standard setups so really don't think the writing off the possibility of it happening is fair. If we want to say people who care about the increase are the minority, then we open up the can of worms again to break/improve bad/existing behavior

**Comment by N-R-K** [comment]
> Simply not true

Just out of curiosity, can you give a single example of (more or less POSIX) OS without virtual memory?

**Comment by TAAPArthur** [comment]
> > Simply not true
> 
> Just out of curiosity, can you give a single example of (more or less POSIX) OS without virtual memory?

You mean without swap? Swap is always optional and one has to take special effort to set it up. On my arch install I never set up a swap partition (think I just forgot) and have been fine. Despite it being incredibly common it is still something one opts into. I can't name a distro that explicitly recommends not having swap as  swapping is generally better than a freeze/crash given one has the disk space. If you want to test one just run `swapoff -a`. From my experience if you don't play intensive games nor use a bloated browser it is kinda hard (at least for me) to use up enough memory to start swapping. So to answer your question, literally any POSIX OS if you want it to.

And just for reference my primary use of an image viewer is as a manga/comic reader where each image is like 200K compressed (don't have the uncompressed numbers off hand). So loading/decompressing the image is negligible and images are viewed in more or less sequentially so non-predictive caching wouldn't help regardless. This change would just make nsxiv use significantly more memory for no gain at least in my usecase. Just sharing a perspective

**Comment by N-R-K** [comment]
> You mean without swap?

No, I'm talking about [virtual memory](https://en.wikipedia.org/wiki/Virtual_memory). I'm aware that swap is optional, I have it disabled myself.

**Comment by TAAPArthur** [comment]
> > You mean without swap?
> 
> No, I'm talking about [virtual memory](https://en.wikipedia.org/wiki/Virtual_memory). I'm aware that swap is optional, I have it disabled myself.

I'm still confused by the question as I don't see the relevance. Not using virtual memory doesn't affect the risk of OOMs. 

>Or I guess a better question would be, which part of that statement is "Simply not true"? The part about virtual memory or ..?

The statement that was quoted? The use of virtual memory is just an abstraction around physical memory and doesn't prevent OOMs itself. Swapping can but the implication that all "modern OS" have it enabled is something I find to false. 

Wasn't trying to say that virtual memory isn't used, but that being conscious of memory usage is still valuable

**Comment by eylles** [comment]
1. I purposefully ignored swap and only had in mind actual physical ram.
2. Would anyone ever use sxiv/nsxiv on a linux phone? Cuz the keyboard driven workflow of sxiv already makes it highly unusable on the mouse heavy touch workflow of a phone.
3. Rasp berry pi zero, i doubt anyon will seriously try to use that as a system with a gui, as a storage server or router yea i have seen those uses, and if someone is to use a pi zero as a NAS then be sure they will mount the remote FS and use local programs.
4. A minimal system with minimal software is all bout not wasting resources on idling so that they can be actually used on the programs that need them, such as image viewers, media players, games, compilers, etc...

**Comment by N-R-K** [comment]
> From my experience if you don't play intensive games nor use a bloated browser it is kinda hard (at least for me) to use up enough memory to start swapping.

In your own post you've described that it's hard to get into swapping, let alone OOM. And it should be a pretty safe assumption that anyone who's on 0.5 gigs is going to be using swap.

So I'm asking how was my statement about it being hard to OOM not true?

**Comment by TAAPArthur** [comment]
>Would anyone ever use sxiv/nsxiv on a linux phone? 

I would. For better or worse nsxiv has mouse support so it is usable out of the box with a touch screen. Then there's also OSKs if someone wanted that full keyboard experience. And to there are also usb keyboards and keyboards as part of the phone case. There are also gestures which can be used to emulate arbitrary key presses.

I don't think we should call this a minimal imagine viewer if there are non-minimal requirements to run it on.

>and if someone is to use a pi zero as a NAS then be sure they will mount the remote FS and use local programs.

A reasonable idea. Using X-forwarding and sshing into the box also seems feasible and is easier to do.

>In your own post you've described that it's hard to get into swapping, 

Hard to get into swapping when software uses memory efficiently (as opposed to bloated browsers). Pretty easy if applications don't free used memory

>And it should be a pretty safe assumption that anyone who's on 0.5 gigs is going to be using swap.

I don't think that assumption is safe. I work comfortably in such an environment without swap. And even if that assumption was safe doesn't mean they would have a lot of swap. And even if they had a lot of swap doesn't mean it is fast. To me swapping is a failure mode and just a politer way of prompting me to close applications than freezing.

Let me take a step back. The purpose of increasing the cache size if perf and I'm trying to make the case that it can break some systems and/or make them slower. I can't tell if the counter argument is that you don't care because they are the minority or don't think they exists. I'm arguing they exists. I can go into detail about specific situations where resources are scarce and applications potentially increasing their memory usage by orders of magnitude can cause problems, hope this is obvious. If we agree that such groups exists, then think maintaining backwards compatibility is enough of a reason to not change the default.

**Comment by N-R-K** [comment]
> The use of virtual memory is just an abstraction around physical memory and doesn't prevent OOMs itself.

It does, using virtual memory as opposed to giving out physical addresses means the kernel can deal with memory fragmentation more optimally. (It also enables swapping.)

> I can go into detail about specific situations where resources are scarce and applications potentially increasing their memory usage by orders of magnitude can cause problems

Sure, go ahead. But keep in mind we're only targeting POSIX systems capable of running X.

---

Also, on the topic of minimalism, sxiv isn't exactly minimal, it's more on the middle. It depends on imlib2, has many features in the code base that couldve been one line shell wrapper and it also doesn't manage memory efficiently either (look into how we're handling options and filenames for example).

**Comment by TAAPArthur** [comment]
>It does, using virtual memory as opposed to giving out physical addresses means the kernel can deal with memory fragmentation more optimally. (It also enables swapping.)

No it allows for solutions that do like swapping. Allows isn't even the right word as swapping could be implemented without VA if one wanted. But what you are describing is an abstraction that gives the kernel more freedom... a means to and end for our discussion which is why I focused on swapping instead of VA.

>Sure, go ahead. But keep in mind we're only targeting POSIX systems capable of running X.

The undertone there was that there should be no need to. And technically we aren't targeting systems that can run X. You can have remote X displays which would have worked with sxiv. Honestly I don't even see the point of this statement as Linux, for example, can run on most non-embedded systems and have already mentioned some with sufficiently low resources.

>Also, on the topic of minimalism, sxiv isn't exactly minimal

You don't have to tell me. It still claims to be minimal though. If there is some minimal system requirement behavior beyond POSIX and having c99 compiler it should stop calling itself minimal/suckless. As in name a another suckless style program that deliberately uses more memory than needed and calls that a feature. I'm not saying nsxiv needs to be as small as possible, but this change is potentially increasing the memory foot print drastically in a way that is harmful to small systems. 

And as far as I can tell, y'all are viewing this as a free and obvious change with no downsides. You need to bump the minimal requirements for that claim to be made.

**Comment by N-R-K** [comment]
> As in name a another suckless style program that deliberately uses more memory than needed and calls that a feature. 

ST uses a font cache.

> y'all are viewing this as a free and obvious change with no downsides

I'm neutral towards changing the default size, which is why I didn't bump the size on my initial PR.

---

But anyhow, if we don't want to make any exceptions as discussed in https://github.com/nsxiv/nsxiv/pull/173#issuecomment-978858667, then there's little point in continuing this.

**Comment by 0ion9** [comment]
if we can count on `sysinfo` being available to get the total amount of ram, then it should be possible to automatically scale cache size to take not more than a given percentage of ram.
Note that I'm not talking about 'available' ram, that's definitely not simple to determine, only total ram.

I don't know whether sysinfo *is* universally available on the more cutdown platforms sxiv supports, for example Rpi. any ideas?

**Comment by N-R-K** [comment]
I've considered having some automatic detection, but I'm not aware of any POSIX function to get the available memory. `sysinfo` is linux specific and we aim for POSIX compatibility.

**Comment by MahouShoujoMivutilde** [comment]
Just wanted to add that 256 MB indeed feels faster than the ancient default when switching between anything remotely hi-res, and it's a real shame binary users will never get to experience that.

Defense for keeping default reads more like defense of some very niche use cases in detriment to majority of users. 

I mean, for real, how many of you actually have primary devices with less than 4 GB of RAM?

Anyway, automatic detection would be the best, but if that's not possible, I think it makes more sense for defaults to be tailored to the most systems and not for _OOM-from-one-browser-tab type of systems_.

If somebody wants to run nsxiv on a system where additional 100MB of RAM usage is a deal breaker - they probably know what they doing, and will run minimal builds of everything that can be optimized anyways.

It's nice that this setting is now out there on display. I didn't knew about it and just assumed that nsxiv is as fast as it can be on my machine. I'm pleasantly surprised that I was proven wrong.

**Comment by eylles** [comment]
on the autodetect front, we can perhaps have it for linux, ask someone with experience with bsd how to get that for bsd and finally for anything else just default to 4 again, i say this considering that suckless software usually just targets linux and bsd support is often patched in.

**Comment by explosion-mental** [comment]
I have to agree on that keeping a 'sane' default here is best, I wil say that this could be changed with `sed` or the like if installing via a package manager.

Those that have a true minimal machine would probably benefit from compiling the programs so editing the config file and lower the cache size isn't a big deal. 

_Used to live on a 2gb intel 2009 (dual core) or something laptop and could open '''many''' tabs on firefox, after switching from xfce4 to dwm experience no freezing._ So I think whether 128 or 256 is the question.

**Comment by N-R-K** [comment]
> So, how much cache is best is the question.

If my calculations are correct then even 32MiB will be a noticeable improvement as it would be able to hold roughly 4 - 1080p images.

The current 4MiB gets filled just by a single 900p image. I wouldn't be surprised if the current default is actually harmful and eating up 4 more MiB needlessly since you're more or less always reloading the image anyways with that small of a cache.

**Comment by MahouShoujoMivutilde** [comment]
> If my calculations are correct then even 32MiB will be a noticeable improvement as it would be able to hold roughly 4 - 1080p images.

Yes, but while 32 MiB will be enough for 1080p, even photos from your normal typical phone camera are already 12-16MP, sometimes larger.

So `24 × 3000 × 4000 × (4 × bit) ≈ 137.3291016 MiB`. 3 photos for 128MiB, 7 for 256MiB. I think 128MiB or preferably 256MiB makes more sense.

**Comment by N-R-K** [comment]
For most desktop usage 128mb should be fine. But that's a bit too much for lower end pi. 64mb seems more reasonable.

Also, while the idea of auto detection is appealing, I do think it's a bit too much for such a simple thing.

> I'm arguing they exists

One thing we can do is bump the size to 64mb after `v28` on git master and add a comment asking users to open a bug report if they're badly affected by this change. That way we will have some time to see how many people, if any, there are who can't afford a +60mb memory.

**Comment by explosion-mental** [comment]
Maybe having this size as a variable (macro) on config.mk, since that is what most distros (package maintainers) would change.

**Comment by TAAPArthur** [comment]
>One other option is to bump the size to 64mb after v28 on git master and add a comment asking users to open a bug report if they're badly affected by this change. That way we will have some time to see how many people there are, if any, who can't afford a +60mb memory.

I feel like affected users would just migrate rather than complain since it looks like people are set on this. I'm just saying that we should stop advertising this a "suckless" image viewer when the target is higher end systems.

But on a general note, doing something and seeing how many people bother to complain is a terrible metric. How many people does it take for it to matter?  Wouldn't they just be told to build from source? If you are just not going to care about minimal systems by default, let's just say that (ideally with recommended minimal specs or something) and not give people surprises. That way you could even bump the number of excessively and not have to worry about fallout.

But honestly thought we agreed to keep defaults to avoid these types of issues.

**Comment by eylles** [comment]
How about an awful terrible idea, open a vote on the default cache size?

**Comment by N-R-K** [comment]
> doing something and seeing how many people bother to complain is a terrible metric.

Just suggesting an option. Also I don't agree that people will switch instead of complaining first nor do I think 64 mb ram is "high end" requirement.

> How about an awful terrible idea, open a vote on the default cache size?

I think the broader issue here is if we want to make exception or not. See: https://github.com/nsxiv/nsxiv/pull/173#issuecomment-978858667 . I voted for not making exceptions, and I haven't changed my mind on it.

Bumping the default size would be nice, but I'm neutral towards it. My main goal was to just expose this setting so people can configure it per their system.

**Comment by MahouShoujoMivutilde** [comment]
>  I'm just saying that we should stop advertising this a "suckless" image viewer when the target is higher end systems

My _phone_ had 2 GB of RAM _back in 2013_, just saying.

There is nothing higher end about such requirement.

And why would "suckless" mean "kind of slow, even though it could have been noticeably faster _at no cost for most users by default_"?

Look, you like to do your thing with extremely low spec machines - go for it. Why should it be default? Considering that you already have your own image viewer. And considering that now you can easily change cache size to suit _your_ needs.

**Comment by N-R-K** [comment]
As I've said before, I don't think Arthur's stance is completely unreasonable given our project scope. So no need to be rude.

---

The reason I suggested to try 64mb out was because I don't think these people exist.
128mb+ I can see that.
64mb _maybe_.
But 32?? I'm all open being proven wrong, but I'm having a hard time imagining someone hand calculating their memory usage and arriving to the point where a mere 32mb increase in usage would lead to OOM.

Which is why the 64mb suggestion, if these people do exist, then we should see one or two of them pop up and we can drop the cache size back down to 4mb (or maybe even disable it completely since at that point it's basically useless).

**Comment by TAAPArthur** [comment]
So I think the people who would have strong opposition to increasing the cache are the minority. And as such the only reasons not to boil down to principle of not changing the defaults or doing what is more convenient for the majority and expecting the minority to adapt. So if such a change was made, the principle lost the argument and there isn't another one to make. I don't see a point of people complaining at that point. How many would need to complain for people be ok losing their increase perf. It is easier to just compile from source and change the default. And this is also why having a cli option/env would be convenient to just save the effort of making trivial changes like this.

>nor do I think 64 mb ram is "high end" requirement.

It isn't a machine with 64MiB, but a machine that couldn't just spare an extra 64MiB in workloads for the arbitrary user.


>The reason I suggested to try 64mb out was because I don't think these people exist. And when I asked for a specific example, I didn't get one.

I don't know what you want for an example. It would affect me. Do you want me to show my full stack and count of memory usage to show how increasing it by 128MiB is problematic? Do you want a list of end hardware that has little memory. I don't think providing those would/has helped because it just seems to be brushed off a niche or unrealistic. I'm describing things that I am/have done and want to be clear that changing the default would be strictly harmful to me and you can see the reaction to that. I'm under the impression that the majority of people who want the increase are thinking of the majority. I'm not knocking that; it just doesn't seem inline with our goals of being compatible with sxiv. The goals could be amended to include something like "for the majority of users" to avoid "niche" counter arguments like mine or just stating what we apparently assume to be a "normal" memory usage.

>Why should it be default?

Because of our goals. I've also suggested other options like making it configurable and/or changing what our goals are.

**Comment by eylles** [comment]
don't you already build from source tho? the default is only useful fro binary users, last time i checked you do modify the config.h and build from source, so why is it a problem that the default isn't what you want if you don't use a pre built binary and you DO modify the defaults anyway?

**Comment by N-R-K** [comment]
> Do you want me to show my full stack and count of memory usage to show how increasing it by 128MiB is problematic?

No, I've said that I can see that happening with 128mb. But a 28mb increase? Can you specify what device you're using where a 28mb increase in memory would be harmful?

**Comment by TAAPArthur** [comment]
Probably, but I agree at that point that the setup is fragile/at its limits. If we went from X to X + epsilon, I'm not trying to say it would be fatal in practice to some people. I just don't see the point. To me the solutions would be to leave it as is out of principle or modify it to suite the average targeted user/majority. A small passive change that seems non-ideal and arbitrary. 

So to answer your question, aside from looking at a memory usage monitor, I doubt I'd notice a 28MiB increase. But given it was increased, I don't know why we'd bother to increase it by so little.

>so why is it a problem that the default isn't what you want

This has been asked a lot and answered as many times. I believe most of us here build from source yet don't believe others have been questioned about "why they care" about what the defaults are. But I'd ask the same to you given you can change it just as easily. I imagine our answers are similar in that it is between users opting in or opting out of this new feature/option. I'm saying they should opt in like they do for basically every other non-bug fix nsxiv has added.

**Comment by N-R-K** [comment]
> But given it was increased, I don't know why we'd bother to increase it by so little.

Because it would be able to hold roughly 4 - 1080p images. The current default of 4Mb holds none. https://github.com/nsxiv/nsxiv/pull/171#issuecomment-979770072

**Comment by TAAPArthur** [comment]
Ok. Why do we care about holding 4 1080p images? Yes a bigger cache holds more images. I get your point that it may be more useful to some, but everyone doesn't frequently have/view 1080p images. Most of the images I view fit into the current cache (sampled the last one I viewed and uncompressed came to 2.5MiB). And why would we favor being able to fit 4*1080p images into the cache over cache memory usage increasing 8x? Why not go to 16MiB instead of 32? I imagine because we are catering to some common usecase/hardware/assumptions. And if we are going to do so, shouldn't we pick an optimal value? I feel the trap of picking a "unoffensive value" like 32MiB is that is is a compromise that makes everyone unhappy.

**Comment by N-R-K** [comment]
> Why not go to 16MiB instead of 32?

That's because a single 1440p image (not uncommon) would take up 14mb space uncompressed, more useful than the current default for people dealing with images of that size, but not too much.

As I see, 32mb is a good common ground between being useful without increasing the mem usage too much (128/256mb) so that smaller devices don't OOM.

**Comment by eylles** [comment]
This is my take, so feel free to disagree and call me names, but all defaults are only useful and will only afect binary users, so the opinions and usecases of the power users and the developers should be deprecated when talking about defaults, why?

* the power users DO change their config.
* thepower users can simply be told "set this default to whatever you like"
* since config is compiletime binary users have no control of it so defaults DO affect them.

So in that light i propose we do deprecate all our personal opinions as devs and power users about what the default cache should be and instead open a vote onlyfor the binary users of nsxiv.

**Comment by N-R-K** [comment]
Having defaults catered toward binary users is fine, but that's not really the point of conflict here. The point of conflict is not increasing the mem usage too much for smaller devices.

For example someone using a pi might very well be a "binary user". In fact, I doubt a pi user would be using a source based distro (unless cross-compiling) since compiling on smaller devices is a pain.

**Comment by 0ion9** [comment]
> Most of the images I view fit into the current cache (sampled the last one I viewed and uncompressed came to 2.5MiB).
If 2.5MiB is representative, then AFAICS the 4MiB cache is doing nothing for you. you need Current Image Size + Previous Image Size, because current image always occupies newest slot in cache.

 I think this needs to be pointed out, as it's not the typical cache behaviour ( 'current thing' doesn't usually count towards cache limit).

**Comment by N-R-K** [comment]
> If 2.5MiB is representative, then AFAICS the 4MiB cache is doing nothing for you. you need Current Image Size + Previous Image Size, because current image always occupies newest slot in cache.

I'm not sure what eviction policy imlib2 uses, but assuming it's LRU then yes. Those 2.5mb are being needlessly wasted, which is why I said it might as well be better to turn the cache off completely at that point.

**Comment by MahouShoujoMivutilde** [comment]
I just googled a bit, and [there seems to be a POSIX way](https://stackoverflow.com/a/2513561) to get RAM size.

The following works on my linux pc and in vagrant freebsd 13 vm.

```c
#include <unistd.h>
#include <stdio.h>

unsigned long long getTotalSystemMemory() {
    long pages = sysconf(_SC_PHYS_PAGES);
    long page_size = sysconf(_SC_PAGE_SIZE);
    return pages * page_size;
}


int main() {
    printf("Total RAM: %llu MB\n", getTotalSystemMemory()/1024/1024);

    return 0;
}
```

**Comment by N-R-K** [comment]
Thanks for the patch. `sysconf` is posix, so is `_SC_PAGE_SIZE`, but `_SC_PHYS_PAGES` unfortunately is not.

And while I'm not against run time detection, I do feel like this is a bit too much for something that should be very simple.

Also a 3% default will end up being around 500mb on a system with 16gigs of memory, and 1gigs for 32gigs system. That's definitely too much for common use cases.

IMO a 32mb default makes most sense. It wont OOM smaller devices but at the same time it will be useful, even for 1440p images.

But if we can't agree on bumping the size statically, then I guess we don't have any other choice but runtime detection.

**Comment by MahouShoujoMivutilde** [comment]
My reasoning for 3% is that it is small enough to not be noticeable, but at the same time big enough to be beneficial for most systems:

* On really tiny systems default 4MiB will be used (maybe should've set this to 0, but idk -https://github.com/nsxiv/nsxiv/pull/171#issuecomment-986164078)
* On limited systems (e.g. 0.5 to 1 gb of RAM) it will be ~15 to ~30 mb, like you wanted.  Still better than current default, but not claiming too much RAM.
* On something typical, which I imagine is probably 4-8gb+, it will be ~125 to ~250 and more. Useful if you use nsxiv to e.g. view photos (13-16MP) from your phone.

My idea is that if you have 16gb of RAM - you probably won't care about 500mb, but will appreciate better performance. At least I would. But we can add the upper limit, like "try to use 3%, but not more than 256mb":

```c
return MIN(MAX(for_cache, 4 * 1024 * 1024), 256 * 1024 * 1024);
```

> but `_SC_PHYS_PAGES` unfortunately is not 

Noticed it now too. Well, that sucks, but according to [sysconf(3)](https://man7.org/linux/man-pages/man3/sysconf.3.html) you can check if variable is supported at runtime:

```c
long long calc_cache_size() {
	long pages = sysconf(_SC_PHYS_PAGES);
	if (pages < 0)
		/* _SC_PHYS_PAGES not supported / error / ..., return whatever we think default should be */
		return 32 * 1024 * 1024;
	/* and so on */
	/* .... */
```

I don't think it is too much: logic is fairly straightforward, and if power user doesn't want it - they'll find this function in no time and override it.

> IMO a 32mb default makes most sense

Frankly, anything is better than 4mb, imo.

**Pull request merged**
